### DISCUSSION TITLE
Sending a message to my Sandbox App from a background job

### DISCUSSION THREAD

**Discussion by randall-coding** [hecrj/iced#937] [Q&A] [answered]
I have my UI set up using iced but I also have jobs running in the background.  I want to update my UI based on something happening in my background job such as if an error occurs.

For example, I spawn my background task using `tokio:spawn()` and when an error occurs in that job I want to call a function from within it to send a message to my running App struct like so `app_message("error message")` which will update the Error object of the App struct and ultimately displays a message onto the screen for the user.  

Does any feature exist to send a message to the App on demand from within my code?  It should still go through the same message pipeline as the UI messages, or if there is another way perhaps that works too.

**Comment by 13r0ck** [comment]
Depends on what exactly you want. Using `tokio::spawn()` is not the intended way of doing async operations with iced, and likely won't work.

If you are looking to start an async operation from you application and then get a response when it finished (succeeds/fails) then a `Command` is what you want. They are the easiest option.

Otherwise you will need to create a subscription. Which are more difficult to implement.

**Reply by randall-coding** [reply]
Thanks.  I will look into using Command or Subscription

**Comment by 13r0ck** [comment]
Also, you can't use a `Sandbox`, you will have to move over to a `Application` for either option.

**Comment by tarkah** [comment, marked as answer]
The best way to implement a background task with `iced` is to implement your task as an  [iced_native::subscription::Recipe](https://docs.rs/iced_native/0.4.0/iced_native/subscription/trait.Recipe.html) and add it as a subscription with [`Application::subcription`](https://docs.rs/iced/0.3.0/iced/trait.Application.html#method.subscription) / [`Subscription::from_recipe`](https://docs.rs/iced_futures/0.3.0/iced_futures/subscription/struct.Subscription.html#method.from_recipe).

This task will get continually polled and you'll be able to send messages directly to your App's `Message` / `update`.

You can see a comprehensive example of how this done for a task that downloads something, and continually sends updates of the download progress here:

https://github.com/hecrj/iced/blob/master/examples/download_progress/src/download.rs

**Reply by randall-coding** [reply]
Thanks that is exactly what I was looking for.
